File: SchoolData/src/main/java/org/gb/homework/schooldata/repository/Repository.java
package org.gb.homework.schooldata.repository;

import org.gb.homework.schooldata.model.User;
import org.gb.homework.schooldata.model.factories.UserFactory;

import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 *
 */
public class Repository<T extends User> {
    private static final boolean isCommitEverytime = true;
    private final String fileDb;
    private final UserFactory<T> userFactory;
    private final Map<Integer, T> users = new HashMap<>();
    private boolean isFetch = false;

    public Repository(String fileDb, UserFactory<T> userFactory) {
        this.fileDb = fileDb;
        this.userFactory = userFactory;
    }

    public void remove(T user) {
        fetch();
        users.get(user.getId()).setActual(false);
        commit();
    }

    public void save(T user) {
        fetch();
        users.put(user.getId(), user);
        commit();
    }

    public List<T> getAll() {
        fetch();
        return new ArrayList<>(users.values());
    }

    public void fetch() {
        if (isFetch)
            return;

        isFetch = true;

        try {
            Path path = Paths.get(fileDb);
            if (Files.exists(path)) {
                List<String> allLines = Files.readAllLines(path);

                for (String line : allLines) {
                    T user = userFactory.deserialize(line);
                    users.put(user.getId(), user);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    public void commit() {
        if (!isCommitEverytime)
            return;

        try (FileWriter fileWriter = new FileWriter(fileDb)) {
            for (var u : users.values()) {
                fileWriter.write(u.serialize());
                fileWriter.write(System.lineSeparator());
                fileWriter.flush();
            }
        } catch (Exception ignored) {
        }
    }
}

File: csharp/SchoolData/Repositories/Repository.cs
using Gb.Homework.SchoolData.Model;
using Gb.Homework.SchoolData.Model.Factories;

namespace Gb.Homework.SchoolData.Repositories
{


    /// <summary>
    /// 
    /// </summary>
    public class Repository<T> : IUserRepository where T : User
    {
        private static readonly bool isCommitEverytime = true;
        private readonly string fileDb;

        private readonly UserFactory<T> userFactory;
        private readonly Dictionary<int, T> users = new();
        private bool isFetch = false;

        public Repository(string fileDb, UserFactory<T> userFactory)
        {
            this.fileDb = fileDb;
            this.userFactory = userFactory;
        }

        public void Remove(T user)
        {
            Fetch();
            users[user.Id].IsActual = false;
            Commit();
        }

        public void Save(T user)
        {
            Fetch();
            users[user.Id] = user;
            Commit();
        }

        public List<T> All
        {
            get
            {
                Fetch();
                return users.Values.ToList();
            }
        }

        List<User> IUserRepository.All => throw new NotImplementedException();

        public void Fetch()
        {
            if (isFetch)
                return;
            isFetch = true;

            try
            {
                if (File.Exists(fileDb))
                {
                    List<String> allLines = File.ReadLines(fileDb).ToList();

                    foreach (String line in allLines)
                    {
                        T user = userFactory.Deserialize(line);
                        users[user.Id] = user;
                    }
                }
            }
            catch (IOException e)
            {
                Console.WriteLine(e.Message);
            }

        }

        public void Commit()
        {
            if (!isCommitEverytime)
                return;

            try
            {
                File.WriteAllLines(fileDb, users.Values.Select(x => x.Serialize()));
            }
            catch (Exception)
            {
            }
        }

        void IUserRepository.Remove(User user)
        {
            Remove((T)user);
        }

        void IUserRepository.Save(User user)
        {
            Save((T)user);
        }
    }
}
